internal/xdscache/v2/endpointstranslator.go

```diff
@@ -298,11 +298,38 @@ func (e *EndpointsTranslator) OnChange(d *dag.DAG) {
 	// set the entries rather than merging them.
 	entries := e.cache.Recalculate()
 
-	e.mu.Lock()
-	e.entries = entries
-	e.mu.Unlock()
+	// Only update and notify if entries has changed.
+	if !equal(e.entries, entries) {
+		e.Debug("cluster load assignments changed, notifying waiters")
+
+		e.mu.Lock()
+		e.entries = entries
+		e.mu.Unlock()
+
+		e.Notify()
+	} else {
+		e.Debug("cluster load assignments did not change")
+	}
+}
+
+// equal returns true if a and b are the same length, have the same set
+// of keys, and have proto-equivalent values for each key, or false otherwise.
+func equal(a, b map[string]*envoy_api_v2.ClusterLoadAssignment) bool {
```

Comment:
I was doing this with `equality.Semantic.DeepEqual` on my branch, but it wasn't working. I can test this out to see how it behaves.

Reply:
Yeah, I observed the same with that or `reflect.DeepEqual`, I think there are some private fields in there that don't match. Doing an item-by-item comparison using `proto.Equal` fixes it, though.

Reply:
Nice!